# Curso: Redes Neurais e Deep Learning



## O Problema a ser Resolvido

O objetivo da atividade é elaborar uma rede neural para predizer se um paciente tem ou não diabetes, com base nas medidas diagnósticas contidas no *dataset* disponível em https://www.kaggle.com/uciml/pima-indians-diabetes-database.

Os dados são de pacientes do sexo feminino, com pelo menos 20 anos de idade. Os atributos das condições médicas incluem o número de gestações que a paciente teve, seu IMC, nível de insulina, idade e outros. A classe a ser predita é o atributo "Outcome", cujos valores são 0 (não tem diabetes) ou 1 (tem diabetes). Portanto, é um problema de classificação binária.

O problema tem como objetivo praticar as seguintes habilidades:
- Efetuar o pré-processamento dos dados, separando-os em conjuntos de treino e teste.
- Configurar uma rede neural simples para um problema de classificação binária.

## Pacotes

In [7]:
import numpy as np # package for scientific computing
import tensorflow as tf  #  package for numerical computation using data flow graphs
from tensorflow import keras  # package for deep learning
import pandas as pd # package for working with structured data

## Pré-Processamentos dos Dados de Treino e de Teste

In [8]:
# Read dataset
data = pd.read_csv('diabetes.csv')
data.head() # display dataset first lines

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1

In [9]:
# Separate the class from other attributes
X = data.drop("Outcome", axis=1)
Y = data["Outcome"]

In [10]:
# Preparing the dataset for training and test
from sklearn.model_selection import train_test_split
train_set_X, test_set_X, train_set_Y, test_set_Y = train_test_split(X, Y, test_size=0.20, random_state=7)

In [11]:
# Standardize features by removing the mean and scaling to unit variance
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

train_set_X = scaler.fit_transform(train_set_X)
test_set_X  = scaler.fit_transform(test_set_X)

In [12]:
# Checking
print("train_set_X:\n", train_set_X[:2,:])
print("\ntest_set_X:\n", test_set_X[:2,:])

train_set_X:
 [[ 0.35483802 -0.370418    0.16624635  1.40032403 -0.01279543  0.49863797
  -0.61786951  0.01064085]
 [-0.54794048 -0.55620696  0.9024924   0.96548604  0.40069886  1.70739891
  -1.0345765  -0.86049025]]

test_set_X:
 [[-0.83168001 -1.14367855 -0.39723824 -0.57849223 -0.36848171 -0.47210798
   0.22800925 -0.84284936]
 [ 0.8647269   1.8633592   0.67258938  0.00976179  0.72124393  0.58740671
   0.24525553  1.28629259]]


In [14]:
n = train_set_X.shape[1]      # number of attributes
m = train_set_X.shape[0]          # number of training examples

print ("Number of attributes: n = " + str(n))
print ("Number of training examples: m = " + str(m))
print ("Train set X shape: " + str(train_set_X.shape))
print ("Train set Y shape: " + str(train_set_Y.shape))
print ("Test set X shape: " + str(test_set_X.shape))
print ("Test set Y shape: " + str(test_set_Y.shape))

Number of attributes: n = 8
Number of training examples: m = 614
Train set X shape: (614, 8)
Train set Y shape: (614,)
Test set X shape: (154, 8)
Test set Y shape: (154,)



---

## Definição do Modelo

Crie um modelo em Keras com a seguinte configuração:
- Camada de entrada: no formato dos dados de entrada do problema
- Camada 1: 3 neurônios, função de ativação *Tanh*
- Camada 2: 5 neurônios, função de ativação *Tanh*
- Camada 3: 3 neurônios, função de ativação *Tanh*
- Camada 4 (saída): 1 neurônio, função de ativação *Sigmoid*


In [15]:
inputs = keras.Input(shape=(train_set_X.shape[1]))  # the shape of inputs (number of attributes)
x = keras.layers.Dense(units=3, activation="tanh")(inputs)
x = keras.layers.Dense(units=5, activation="tanh")(x)
x = keras.layers.Dense(units=3, activation="tanh")(x)
outputs = keras.layers.Dense(units=1, activation="sigmoid")(x)  # single neuron network with sigmoid as activation function
model = keras.Model(inputs=inputs, outputs=outputs) 

In [16]:
# Checking
processed_data = model(train_set_X)
print(processed_data.shape)

(614, 1)


**SAÍDA ESPERADA**: <br>
(614, 1)

---

In [17]:
# Prints a summary of the network, showing its architecture and parameters.
model.summary()

Model: "model"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 8)]               0         
                                                                 
 dense (Dense)               (None, 3)                 27        
                                                                 
 dense_1 (Dense)             (None, 5)                 20        
                                                                 
 dense_2 (Dense)             (None, 3)                 18        
                                                                 
 dense_3 (Dense)             (None, 1)                 4         
                                                                 
Total params: 69
Trainable params: 69
Non-trainable params: 0
_________________________________________________________________


## Compilação do Modelo

In [18]:
model.compile(optimizer="RMSprop", loss="mean_absolute_error", metrics=["accuracy","Precision","Recall"])

## Treinamento do Modelo

In [19]:
history = model.fit(train_set_X, train_set_Y, batch_size=64, epochs=1000) 
print(history.history)  # print per-epoch timeseries of metrics values

Epoch 1/1000
10/10 [==============================] - 1s 2ms/step - loss: 0.4897 - accuracy: 0.5472 - precision: 0.3755 - recall: 0.4787
Epoch 2/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.4788 - accuracy: 0.5684 - precision: 0.3826 - recall: 0.4171
Epoch 3/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.4698 - accuracy: 0.5847 - precision: 0.3922 - recall: 0.3791
Epoch 4/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.4609 - accuracy: 0.6010 - precision: 0.4096 - recall: 0.3649
Epoch 5/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.4525 - accuracy: 0.6140 - precision: 0.4253 - recall: 0.3507
Epoch 6/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.4439 - accuracy: 0.6287 - precision: 0.4459 - recall: 0.3318
Epoch 7/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.4355 - accuracy: 0.6417 - precision: 0.4681 - recall: 0.3128
Epoch 8/1000
10/10 [=====================

10/10 [==============================] - 0s 2ms/step - loss: 0.2530 - accuracy: 0.7622 - precision: 0.6836 - recall: 0.5735
Epoch 61/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.2520 - accuracy: 0.7622 - precision: 0.6816 - recall: 0.5782
Epoch 62/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.2512 - accuracy: 0.7622 - precision: 0.6816 - recall: 0.5782
Epoch 63/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2503 - accuracy: 0.7638 - precision: 0.6833 - recall: 0.5829
Epoch 64/1000
10/10 [==============================] - ETA: 0s - loss: 0.2197 - accuracy: 0.7969 - precision: 0.6667 - recall: 0.631 - 0s 2ms/step - loss: 0.2496 - accuracy: 0.7655 - precision: 0.6872 - recall: 0.5829
Epoch 65/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2487 - accuracy: 0.7655 - precision: 0.6872 - recall: 0.5829
Epoch 66/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2480 - accuracy: 0.7638 - pre

10/10 [==============================] - 0s 1ms/step - loss: 0.2214 - accuracy: 0.7850 - precision: 0.7337 - recall: 0.5877
Epoch 119/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2207 - accuracy: 0.7834 - precision: 0.7294 - recall: 0.5877
Epoch 120/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2199 - accuracy: 0.7883 - precision: 0.7368 - recall: 0.5972
Epoch 121/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2195 - accuracy: 0.7899 - precision: 0.7412 - recall: 0.5972
Epoch 122/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2193 - accuracy: 0.7915 - precision: 0.7427 - recall: 0.6019
Epoch 123/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.2188 - accuracy: 0.7932 - precision: 0.7471 - recall: 0.6019
Epoch 124/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2187 - accuracy: 0.7932 - precision: 0.7471 - recall: 0.6019
Epoch 125/1000
10/10 [====================

10/10 [==============================] - 0s 1ms/step - loss: 0.2067 - accuracy: 0.7997 - precision: 0.7558 - recall: 0.6161
Epoch 178/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.2064 - accuracy: 0.8013 - precision: 0.7572 - recall: 0.6209
Epoch 179/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.2063 - accuracy: 0.7997 - precision: 0.7529 - recall: 0.6209
Epoch 180/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2063 - accuracy: 0.7997 - precision: 0.7529 - recall: 0.6209
Epoch 181/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2061 - accuracy: 0.8013 - precision: 0.7572 - recall: 0.6209
Epoch 182/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.2060 - accuracy: 0.8013 - precision: 0.7572 - recall: 0.6209
Epoch 183/1000
10/10 [==============================] - ETA: 0s - loss: 0.1052 - accuracy: 0.9062 - precision: 0.9375 - recall: 0.750 - 0s 1ms/step - loss: 0.2057 - accuracy: 0.8013

10/10 [==============================] - 0s 1ms/step - loss: 0.1988 - accuracy: 0.8062 - precision: 0.7706 - recall: 0.6209
Epoch 236/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1985 - accuracy: 0.8062 - precision: 0.7706 - recall: 0.6209
Epoch 237/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1988 - accuracy: 0.8062 - precision: 0.7706 - recall: 0.6209
Epoch 238/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1984 - accuracy: 0.8046 - precision: 0.7692 - recall: 0.6161
Epoch 239/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1983 - accuracy: 0.8062 - precision: 0.7706 - recall: 0.6209
Epoch 240/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1983 - accuracy: 0.8062 - precision: 0.7706 - recall: 0.6209
Epoch 241/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1980 - accuracy: 0.8062 - precision: 0.7706 - recall: 0.6209
Epoch 242/1000
10/10 [====================

10/10 [==============================] - 0s 1ms/step - loss: 0.1946 - accuracy: 0.8078 - precision: 0.7719 - recall: 0.6256
Epoch 295/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.1944 - accuracy: 0.8078 - precision: 0.7719 - recall: 0.6256
Epoch 296/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1943 - accuracy: 0.8078 - precision: 0.7719 - recall: 0.6256
Epoch 297/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1944 - accuracy: 0.8078 - precision: 0.7719 - recall: 0.6256
Epoch 298/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1944 - accuracy: 0.8078 - precision: 0.7719 - recall: 0.6256
Epoch 299/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1943 - accuracy: 0.8078 - precision: 0.7719 - recall: 0.6256
Epoch 300/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1943 - accuracy: 0.8078 - precision: 0.7719 - recall: 0.6256
Epoch 301/1000
10/10 [====================

10/10 [==============================] - 0s 2ms/step - loss: 0.1922 - accuracy: 0.8094 - precision: 0.7733 - recall: 0.6303
Epoch 354/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1918 - accuracy: 0.8094 - precision: 0.7733 - recall: 0.6303
Epoch 355/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1917 - accuracy: 0.8094 - precision: 0.7733 - recall: 0.6303
Epoch 356/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.1919 - accuracy: 0.8094 - precision: 0.7733 - recall: 0.6303
Epoch 357/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1918 - accuracy: 0.8094 - precision: 0.7733 - recall: 0.6303
Epoch 358/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.1918 - accuracy: 0.8094 - precision: 0.7733 - recall: 0.6303
Epoch 359/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.1917 - accuracy: 0.8094 - precision: 0.7733 - recall: 0.6303
Epoch 360/1000
10/10 [====================

10/10 [==============================] - 0s 1ms/step - loss: 0.1897 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 413/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1898 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 414/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1898 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 415/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1896 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 416/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1896 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 417/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1896 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 418/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1897 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 419/1000
10/10 [====================

10/10 [==============================] - 0s 1ms/step - loss: 0.1892 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 472/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1892 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 473/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1891 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 474/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1891 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 475/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1891 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 476/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1892 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 477/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1891 - accuracy: 0.8111 - precision: 0.7778 - recall: 0.6303
Epoch 478/1000
10/10 [====================

10/10 [==============================] - 0s 2ms/step - loss: 0.1884 - accuracy: 0.8127 - precision: 0.7824 - recall: 0.6303
Epoch 531/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1879 - accuracy: 0.8127 - precision: 0.7824 - recall: 0.6303
Epoch 532/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1882 - accuracy: 0.8127 - precision: 0.7824 - recall: 0.6303
Epoch 533/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1882 - accuracy: 0.8127 - precision: 0.7824 - recall: 0.6303
Epoch 534/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1877 - accuracy: 0.8127 - precision: 0.7824 - recall: 0.6303
Epoch 535/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1878 - accuracy: 0.8127 - precision: 0.7824 - recall: 0.6303
Epoch 536/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1882 - accuracy: 0.8127 - precision: 0.7824 - recall: 0.6303
Epoch 537/1000
10/10 [====================

10/10 [==============================] - 0s 1ms/step - loss: 0.1858 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 590/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1858 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 591/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1858 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 592/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1858 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 593/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1858 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 594/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1858 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 595/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1858 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 596/1000
10/10 [====================

10/10 [==============================] - 0s 1ms/step - loss: 0.1858 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 648/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 649/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 650/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 651/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1858 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 652/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 653/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 654/1000
10/10 [====================

10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 706/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 707/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 708/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 709/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 710/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 711/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 712/1000
10/10 [====================

10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 764/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 765/1000
10/10 [==============================] - ETA: 0s - loss: 0.2188 - accuracy: 0.7812 - precision: 0.6923 - recall: 0.473 - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 766/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 767/1000
10/10 [==============================] - 0s 997us/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 768/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 769/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.81

10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 822/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 823/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 824/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 825/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 826/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 827/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 828/1000
10/10 [====================

10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 881/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 882/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 883/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 884/1000
10/10 [==============================] - 0s 2ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 885/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 886/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 887/1000
10/10 [====================

10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 939/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 940/1000
10/10 [==============================] - 0s 997us/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 941/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 942/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 943/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 944/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 945/1000
10/10 [==================

10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 997/1000
10/10 [==============================] - 0s 997us/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 998/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 999/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
Epoch 1000/1000
10/10 [==============================] - 0s 1ms/step - loss: 0.1857 - accuracy: 0.8143 - precision: 0.7870 - recall: 0.6303
{'loss': [0.4897180497646332, 0.478820264339447, 0.46981343626976013, 0.4609294831752777, 0.45246922969818115, 0.4438782334327698, 0.435464084148407, 0.4267793893814087, 0.418473482131958, 0.4104032814502716, 0.4027360677719116, 0.39540258049964905, 0.38832637667655945, 0.3817918598651886, 0.3755265474319458, 0.3697

**SAÍDA ESPERADA**: <br>
Na época 1000, tem-se o seguinte resultado (aproximado): <br>
loss: 0.1938 - accuracy: 0.8062 - precision: 0.7644 - recall: 0.6303

---

## Avaliação do Modelo


In [20]:
loss, acc, prec, rec = model.evaluate(test_set_X, test_set_Y)
print("Loss: %.2f" % loss, "\nAccuracy: %.2f" % acc, "\nPrecision: %.2f" % prec, "\nRecall: %.2f" % rec)

5/5 [==============================] - 0s 2ms/step - loss: 0.2058 - accuracy: 0.7922 - precision: 0.7660 - recall: 0.6316
Loss: 0.21 
Accuracy: 0.79 
Precision: 0.77 
Recall: 0.63


**SAÍDA ESPERADA**: <br>
Valores aproximados:

Loss: 0.21  <br>
Accuracy: 0.79  <br>
Precision: 0.76  <br>
Recall: 0.61

---

## Predição

In [21]:
predictions = model.predict(test_set_X)
print("Predictions: ", [round(x[0]) for x in predictions])
print("\nCorrect:     ", [round(x) for x in test_set_Y])

Predictions:  [0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0]

Correct:      [0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0]


**SAÍDA ESPERADA**: <br>
Valores aproximados: <br>
Predictions:  [0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0]

Correct:      [0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0]

---